src/foam/nanos/auth/UserAndGroupAuthService.java

```diff
@@ -129,10 +148,26 @@ public User login(X x, long userId, String password) throws AuthenticationExcept
     Session session = x.get(Session.class);
     session.setUserId(user.getId());
     session.setContext(session.getContext().put("user", user));
+    session.setContext(session.getContext().put("group", group));
```

Comment:
I don't understand why we're putting a group in the context. Everywhere that looks at the group looks at the group on the user in the context.

Reply:
Yeah. Doing this is going to introduce a lot of confusion and improper code.

Reply:
That could work but I still think they should be separated considering the new agent user relationship group isn't part of the user. It's part of a relationship junction object. And I guess the user in the context isn't necessarily the user either but consider what we did with our bareUserDAO, where we striped all authentication on the DAO. All new DAOs storing users will have to account for the limiting condition. It sort of becomes a hidden property value in my opinion, I'm kinda iffy on it.

Reply:
> All new DAOs storing users will have to account for the limiting condition.

The same decorator could be reused on any new DAOs that store `User`s and are separate from the existing `userDAO` and `bareUserDAO`.

---

I think keeping group on the user only introduces fewer changes to the system and less complexity than adding a group to the context and is therefore the better solution. The issue of an agent updating a super user's group is important, but easily mitigated by a simple decorator that can be applied to any DAO that needs it, and right now there's only one DAO that needs it anyway, making things very simple.

Reply:
> I think keeping group on the user only introduces fewer changes to the system and less complexity than adding a group to the context and is therefore the better solution.

I don't see how having a group exist within the context is complex. The changes required to make this work is minimal and accounted for within this PR (hopefully) The thing I see with this being a problem is other applications using FOAM accounting for this change but the same thing goes for what you're purposing.

I think that once the group context is introduced, it will lead to less problems in the future.

Reply:
> I don't see how having a group exist within the context is complex.

Here's how it's more complex. Currently the group is only in one place: on the user. What you're proposing is having the group be in two different places. That's more complex because any code that wants to look at a User's group has to know which group to look at. If the group were only in one place like it currently is it would be unambiguous.

> The changes required to make this work is minimal and accounted for within this PR (hopefully)

There are three methods in this file alone - `auth.userAndGroupContext(), auth.checkUserPermission()` and `auth.updatePassword()` - that weren't accounted for. Who knows how many other places were missed? Who knows how many places in the future are going to be missed? And even if you did change them all, why add the extra complexity when it's not necessary?

> The thing I see with this being a problem is other applications using FOAM accounting for this change but the same thing goes for what you're purposing.

You're correct that both cases require change. Here are some potential solutions:

1. Don't put this functionality in FOAM. We only need it in Ablii at the moment and there's no other use case for it that I've heard of. As a rule of thumb I disagree with making a general solution to a specific problem unless you have at least two or preferably three use cases that would justify the extra level of abstraction.
2. Keep it in FOAM, but add the decorator I proposed to the `userDAO` in FOAM by default.

> I think that once the group context is introduced, it will lead to less problems in the future.

Why do you think having the group in the context will lead to less problems in the future?